File: core/src/com/broll/gainea/server/core/objects/buffs/BuffableValue.java
package com.broll.gainea.server.core.objects.buffs;

import java.util.HashSet;
import java.util.Set;

public abstract class BuffableValue<T, B extends Buff, V> {

    private T object;
    protected V value;
    protected V buffedValue;
    protected Set<B> buffs = new HashSet<>();

    public BuffableValue(T object) {
        this.object = object;
    }

    public void addBuff(B buff) {
        this.buffs.add(buff);
        buff.register(this);
        recalc();
    }

    void clearBuff(B buff) {
        this.buffs.remove(buff);
        recalc();
    }

    protected abstract void recalc();

    public void clearBuffs() {
        buffs.clear();
    }

    public T getObject() {
        return object;
    }

    public void setValue(V value) {
        this.value = value;
        recalc();
    }

    public V getValue() {
        return buffedValue;
    }

    public V getRootValue() {
        return value;
    }

    @Override
    public String toString() {
        return ""+buffedValue;
    }
}

File: server/src/main/com/broll/gainea/server/core/objects/buffs/BuffableValue.kt
package com.broll.gainea.server.core.objects.buffs

abstract class BuffableValue<T, V, B : Buff<V>>(val target: T, initValue: V) {

    private var frozen = false
    var value: V
        set(v) {
            if (frozen) return
            rootValue = v
            recalc()
        }
        get() = buffedValue

    var rootValue: V = initValue
    protected var buffedValue: V = initValue
    protected var buffs = mutableSetOf<B>()
    fun addBuff(buff: B) {
        if (frozen) return
        buffs.add(buff)
        buff.register(this as BuffableValue<*, V, Buff<V>>)
        recalc()
    }

    fun clearBuff(buff: B) {
        if (frozen) return
        buffs.remove(buff)
        recalc()
    }

    fun freeze() {
        frozen = true
    }

    protected abstract fun recalc()
    fun clearBuffs() {
        if (frozen) return
        buffs.clear()
    }

    override fun toString() = value.toString()
}
